### PR TITLE
tt-train: one-step async GRPO trainer + phase-split refactor + MeshSocketWeightBridge

### PR DESCRIPTION
**Pull request by ichovpanTT**
### Summary

Stacked on #54390 (gsm8k example). Three related additions on the `tt-train` async-RL path:

* **`OneStepAsyncGRPOTrainer`** — verl-shape one-step off-policy GRPO trainer that overlaps generation of batch `t+1` on the remote rollout worker with the optimizer step on batch `t`. Staleness is exactly 1: rollouts `R_t` were generated with `theta_{t-1}` at the moment we consume them against actor `theta_t` to produce `theta_{t+1}`. Live example wiring at `tt-train/sources/examples/grpo_remote_rollout/gsm8k_onestep/`.
* **`GRPOTrainer` phase-split refactor** — `train()` now walks one method per phase (`_iter_prompt_batches` → `_rollout` / `_await_rollout` → `_expand_prompts_and_columns` → `_compute_rewards` → `_compute_advantages` → `_optimize` → `_apply_gradients` → `_publish_step_metrics` → `_maybe_checkpoint` → `_reset_step_metrics`). Every phase helper takes 1-3 raw args and writes metrics directly into `self.metrics`, replacing the previous 5-8 argument signatures. Killed side-channels: `self._reward_component_means`, `self._callback_times`, `warmup_factor` plumbing. `OneStepAsyncGRPOTrainer` now overrides only `train()` for the prime/await/resubmit loop shape; every phase helper is inherited.
* **`MeshSocketWeightBridge`** — fabric-based alternative to `HostWeightBridge` for the ttml -> ttt weight-sync path. On Qwen3-0.6B: ~0.85 s per push vs ~3.9 s over MPI (~4.6x speedup); see [`WEIGHT_SYNC_OPTIMIZATIONS.md`](tt-train/sources/examples/grpo_remote_rollout/gsm8k/WEIGHT_SYNC_OPTIMIZATIONS.md) for the design + further optimization stack. Selectable via `weight_bridge: {host | mesh_socket}` in the gsm8k example config.

### Notes for reviewers

* Base is `ichovpan/gsm8k-example-async-rl` (open as #54390). Once that lands this can rebase to main.
* Test coverage: device-less loop-shape tests for `OneStepAsyncGRPOTrainer` (prime/await/resubmit interleaving, `num_iterations == 1` guard, completer-contract preflight, `_publish_step_metrics` splat regression) at `tt-train/tests/python/test_one_step_async_grpo_trainer.py`. `MeshSocketWeightBridge` 1->1 transport test at `tt-train/tests/python/grpo_remote_rollout/weight_transfer/`.
* One `fabric` change carried along: drop `/2` on the `num_assigned` intermesh check to fix the 2-mesh single-cable case that the MeshSocketWeightBridge topology triggers.

Made with Cursor

Made with [Cursor](https://cursor.com)

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ichovpan/tt-train-1-step-async-rl)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ichovpan/tt-train-1-step-async-rl)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ichovpan/tt-train-1-step-async-rl)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ichovpan/tt-train-1-step-async-rl)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ichovpan/tt-train-1-step-async-rl)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ichovpan/tt-train-1-step-async-rl)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ichovpan/tt-train-1-step-async-rl)

### CODE REVIEW

File: tt-train/sources/examples/grpo_remote_rollout/gsm8k/gsm8k_training_example.py [hunk 1]
```diff
@@ -226,7 +253,8 @@ def _ttml_main() -> None:
     completer: Any = None
     client: Any = None
     try:
-        bridge = HostWeightBridge.init_sender(mesh=mesh_device, peer_rank=TTT_RANK)
+        weight_bridge_kind = _resolve_weight_bridge_kind(raw)
```
Comment (jmalone-tt) [inline tt-train/sources/examples/grpo_remote_rollout/gsm8k/gsm8k_training_example.py:256]: can simplify - we don't need double validation here or for the receiver bridge creation

File: tt-train/sources/examples/grpo_remote_rollout/gsm8k/plot_rewards.py [hunk 2]
```diff
@@ -5,11 +5,10 @@
```
Comment (jmalone-tt) [inline tt-train/sources/examples/grpo_remote_rollout/gsm8k/plot_rewards.py:5]: can you add a comment to the PR description that shows how plots are changed (old vs new)

File: tt-train/sources/examples/grpo_remote_rollout/gsm8k/WEIGHT_SYNC_OPTIMIZATIONS.md [hunk 3]
```diff
@@ -0,0 +1,260 @@
+# Cutting `MeshSocketWeightBridge` sync time toward 0.5 s
```
Comment (jmalone-tt) [inline tt-train/sources/examples/grpo_remote_rollout/gsm8k/WEIGHT_SYNC_OPTIMIZATIONS.md:1]: don't need to commit this file

File: tt-train/sources/examples/grpo_remote_rollout/gsm8k_onestep/gsm8k_onestep_training_example.py [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+        str(REPO_ROOT),
+        "generated/tt-train/grpo_gsm8k_onestep_run",
+        datetime.now(timezone.utc).strftime("%Y%m%d_%H%M%S"),
+    )
+
+
+_VALID_WEIGHT_BRIDGES = ("host", "mesh_socket")
+
+
+def _resolve_weight_bridge_kind(raw: dict) -> str:
```
Comment (jmalone-tt) [inline tt-train/sources/examples/grpo_remote_rollout/gsm8k_onestep/gsm8k_onestep_training_example.py:192]: This doesn't seem like it belongs in the example (should be an implementation detail the user doesn't need to worry about other than selecting which to use)

File: tt-train/sources/examples/grpo_remote_rollout/gsm8k_onestep/gsm8k_onestep_training_example.py [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+            inference_client=client,
+            enable_ddp=device_config.enable_ddp,
+        )
+
+        # Async trainer owns all weight pushes -- no eager boot push, no
+        # WeightSyncCallback. The prime call inside train() pushes theta_0
+        # before submitting gen_0.
+        client.connect()
+
+        trainer = OneStepAsyncGRPOTrainer(
```
Comment (jmalone-tt) [inline tt-train/sources/examples/grpo_remote_rollout/gsm8k_onestep/gsm8k_onestep_training_example.py:283]: Maybe this is the right place to pass the desired weight bridge and GRPOTrainer can set everything up, with a default value that will be used if unspecified.

Comment (jmalone-tt) [reply tt-train/sources/examples/grpo_remote_rollout/gsm8k_onestep/gsm8k_onestep_training_example.py:283]: I think having separate classes here also isn't the best design. Why have separate OneStepAsyncGRPOTrainer and GRPOTrainer instead of a unified GRPOTrainer that can handle both?

File: tt-train/tests/python/grpo_remote_rollout/weight_bridge/configurations/1_1/mgd.textproto [hunk 6]
```diff
@@ -0,0 +1,44 @@
+# Mesh Graph Descriptor for the MeshSocketWeightBridge 1->1 transport test.
```
Comment (jmalone-tt) [inline tt-train/tests/python/grpo_remote_rollout/weight_bridge/configurations/1_1/mgd.textproto:1]: Fine to keep, but the weight bridge changes and test probably could have been a separate PR so reviewing would have been easier

File: tt-train/tests/python/grpo_remote_rollout/weight_transfer/test_weight_transfer.py [hunk 7]
```diff
@@ -61,8 +61,12 @@
 NUM_SUBMESHES = 4
 
 # Post-push RPC batch; TTT_MAX_BATCH_SIZE must be >= this.
-POST_PUSH_BATCH = 8
-TTT_MAX_BATCH_SIZE = 8
+# Bumped from 8 -> 16: TttGenerationWorker on Blackhole exhibits a determinism
+# regression at max_batch_size=8 (some slots diverge from the greedy output);
+# running with max_batch_size=16 works around it until the underlying worker
+# bug is fixed.
```
Comment (jmalone-tt) [inline tt-train/tests/python/grpo_remote_rollout/weight_transfer/test_weight_transfer.py:67]: do we have a tracking ticket for this issue?

File: tt-train/sources/ttml/ttml/trainers/grpo_trainer.py [hunk 8]
```diff
@@ -1105,194 +1086,353 @@ def train(self) -> None:
...
+        self,
+        prompts_x: List[List[int]],
+        completions: List[List[int]],
+        advantages_np: np.ndarray,
+    ) -> None:
+        """Reference log-probs (once) + per-micro-batch forward-with-grad + GRPO
+        loss + backward. Accumulates gradients across the whole generation
+        batch; does NOT call ``optimizer.step()`` — that is ``_apply_gradients``.
+        """
+        ref_logprobs = self._compute_ref_logprobs(prompts_x, completions)
```
Comment (jmalone-tt) [inline tt-train/sources/ttml/ttml/trainers/grpo_trainer.py:1216]: I think _optimize is called once per mini-epoch, so the ref_logprobs will change with each mini-epoch, which doesn't match the policy they were computed with. This will be an issue when we use a config with num_iterations > 1. Takeaway here is that ref_logprobs should be calculated once per generation and then passed into _optimize

File: tt-train/sources/ttml/ttml/trainers/grpo_trainer.py [hunk 9]
```diff
@@ -1105,194 +1086,353 @@ def train(self) -> None:
...
+            model_source=self.model_source,
+        )
+        ckpt_dir = os.path.join(cfg.output_dir, "checkpoints", f"grpo_step_{step}")
+        monitor_cb = next((cb for cb in self.callbacks if isinstance(cb, GRPOMonitor)), None)
+        for cb in self.callbacks:
+            if cb is monitor_cb:
+                continue
+            self._time_callback(cb, "on_save", self, step, ckpt_dir)
+
+    def _reset_step_metrics(self) -> None:
```
Comment (jmalone-tt) [inline tt-train/sources/ttml/ttml/trainers/grpo_trainer.py:1397]: Looks like this is called inside mini-epoch loop which will wipe batch-level reward metrics when num_iterations > 1

File: tt_metal/fabric/control_plane.cpp [hunk 10]
```diff
@@ -2769,9 +2769,14 @@ void ControlPlane::generate_intermesh_connectivity() {
...
-    // Divide by 2 here, since the intermesh_connections data structure stores connections
-    // bidirectionally.
-    auto num_assigned_intermesh_connections = intermesh_connections.size() / 2;
+
+    // intermesh_connections stores each cable bidirectionally (one directed slot per direction).
+    // We compare directed-slot counts on both sides of the check below: requested_intermesh_connections
+    // is likewise populated per direction by the MGD parser, so the raw .size()s are in the same unit.
+    // Do NOT divide by 2 here -- that historical /2 (introduced in PR #34397) put the two sides in
+    // mismatched units and made any 2-mesh, single-cable topology unsatisfiable even though one
+    // physical cable is topologically sufficient for a mesh pair.
```
Comment (jmalone-tt) [inline tt_metal/fabric/control_plane.cpp:2778]: I don't think this change is right - can you double check if things resolve without it?
From claude:
requested_intermesh_connections_ is one-directional, and the getter returns the outer map size - source meshes, not directed channels. Both producers of intermesh_connections emit two entries per cable, so the /2 was correct.

File: tt-train/tests/python/grpo_remote_rollout/weight_transfer/test_weight_transfer.py (hunk 7 above)
Comment (ichovpanTT) [reply tt-train/tests/python/grpo_remote_rollout/weight_transfer/test_weight_transfer.py:67]: Just submitted, https://github.com/tenstorrent/tt-metal/issues/55425

**Comment by ichovpanTT** [comment]
Having fully async trainer is the priority for now, closing the PR
